Enforce either optional or required tag on apiserverinternal API group

JoelSpeed opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind api-change

#### What this PR does / why we need it:

This PR is the first in a wave of enabling enforcement that every API field should be marked either `+optional` XOR `+required`.

Starting with the `apiserverinternal` API group. I've marked each field that previously wasn't marked with optionality based on reviewing the storage implementation and the behaviour when the field is unspecified.

All other API groups are ignored for now (as there are another ~900 issues to fix), with the plan being to enable these groups 1:1 moving forward.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->
Relates to #134671

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Updates API server internal API group to improve openapi schema correctness for fields being optional or required
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

JoelSpeed committed:
Enforce either optional or required tag on apiserverinternal API group

JoelSpeed committed:
Remove exceptions for groups with no issues

JoelSpeed commented on hack/kube-api-linter/exceptions.yaml:
```diff
@@ -19,3 +19,11 @@
+
+# OptionalOrRequired is being enabled over time. For now, each API group should be added to this list until we comb through each group and fix the missing tags.
+- text: "must be marked as optional or required"
+  path: "staging/src/k8s.io/api/(admission|apidiscovery|apps|authentication|authorization|autoscaling|batch|certificates|core|discovery|events|extensions|flowcontrol|imagepolicy|networking|node|policy|rbac|resource|scheduling|storage)"
```
Will remove these 1x1 when we have the time to tackle the issues

Enumerated the count per group over https://github.com/kubernetes/kubernetes/issues/134671#issuecomment-3415486017

JoelSpeed commented on staging/src/k8s.io/api/apiserverinternal/v1alpha1/types.go:
```diff
@@ -67,20 +70,24 @@ type StorageVersionStatus struct {
 	// The ID of the reporting API server.
-	APIServerID string `json:"apiServerID,omitempty" protobuf:"bytes,1,opt,name=apiServerID"`
+	// +required
+	APIServerID string `json:"apiServerID" protobuf:"bytes,1,opt,name=apiServerID"`
```
Validated as DNSSubdomain, empty string is an error

aaron-prindle commented:
This PR needs `hack/update-codegen.sh` to be run with the updated +optional and +required tags (currently CI failing due to this)

liggitt commented on hack/kube-api-linter/exceptions.yaml:
```diff
@@ -19,3 +19,11 @@
+
+# OptionalOrRequired - Existing fields that are marked as both optional and required (based on standard optional vs kubebuilder:validation:Required) and should not be fixed.
+- text: "field Error must not be marked as both optional and required"
+  path: "staging/src/k8s.io/api/(core/v1|extensions/v1beta1|networking/(v1|v1beta1))"
```
Isn't marking this field as required in kubebuilder incorrect? The API docs on `Ports` say every port in the service should have an entry in the list, which means if a single port has an error they want to report, all ports must, even ones which don't have errors... so... the error field is optional.

I'd expect to resolve this by dropping the kubebuilder Required tag.

```
	// Ports is a list of records of service ports
	// If used, every port defined in the service should have an entry in it
	// +listType=atomic
	// +optional
	Ports []PortStatus `json:"ports,omitempty" protobuf:"bytes,4,rep,name=ports"`
```

thockin replied:
```
Error *string `json:"error,omitempty" protobuf:"bytes,3,opt,name=error"`
```

Won't omitempty omit `nil` but still treat `""` as "has a value" ?

JoelSpeed replied:
> Won't omitempty omit nil but still treat "" as "has a value" ?

I missed that this was a pointer, but with Jordan's point, still stands that a non empty error is required

thockin commented on staging/src/k8s.io/api/apiserverinternal/v1alpha1/types.go:
```diff
@@ -28,13 +28,16 @@ import (
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
 	// Spec is an empty spec. It is here to comply with Kubernetes API style.
+	// +optional
```
WRT https://docs.google.com/document/d/1xpr32B3EMH8LPROj5-TxCUcsABShf_PbIZNy4EmFMYM/edit?tab=t.0

Declarative validation is trying to drive logic from these tags.  To say that a non-pointer struct is optional for an particular operation makes sense but unless we change the Go types that we expose through clients, it can really only mean "if this struct is equivalent to the zero-value, treat it as not-specified".   That seems OK but it leads to the inconsistency cited as "part one" in that doc.

Are we OK with that?  I don't think so.  Assuming your mission is successful, we will end up with a lot non-pointer structs marked as optional.  What does it means for a type's `.Spec` to be optional?  We don't want to use the zero-value check to bypass validation, because .spec isn't schematically optional, but the serialized payload for a given API operation might omit it (e.g. a PUT to /status)

I still think it could be useful (for builtin types) to forbid the use of a `required` field in a non-pointer struct field which is itself marked `optional`.  Likewise, if a non-pointer struct field is required but has no required children, the "is equivalent to the zero-value" logic doesn't work.

Those would remove ambiguity by simply not allowing it, but we need to answer the above. 

Very concretely: Is `Pod.Spec` optional or required?  For some operations it is not required to be present but it must exist in the stored data.

JoelSpeed replied:
> Declarative validation is trying to drive logic from these tags. To say that a non-pointer struct is optional for an particular operation makes sense but unless we change the Go types that we expose through clients, it can really only mean "if this struct is equivalent to the zero-value, treat it as not-specified". That seems OK but it leads to the inconsistency cited as "part one" in that doc.

So I'm running on the assumption we cannot change these structs to pointers as the APIs have shipped. Which means that to be accurate about the validation of optionality, we have to change the way we are validating optionality.

I would pose that the definition of optionality we want is based on the presence of the json key. If we mark a field required, the json key must exist in the serialised data and then we apply validation. If we mark a field optional, then the key may or may not exist in the serialised data, and as such, we can skip internal validation if the key is not present.

I am not fluent in proto or the DV work, or the limitations that they impose upon us, but, have we looked into ways that we could use a serialised presence check to implement optionality?

thockin replied:
Changing the structs would be akin to making a v2 of the client -- we still have to cope with v1.

> I would pose that the definition of optionality we want is based on the presence of the json key.

That's a possible interpretation, but I'm dubious.  We know that JSON and proto can serialize the same Go structs differently, ESPECIALLY around non-pointer structs.  Are we saying that the JSON serialization is the canonically correct one?  Aside from proto not conforming to that, we don't even use `omitzero` so I am not sure that definition works.  Let's look at a trivial example:

https://go.dev/play/p/pnprJcsm1qc

The result of JSON serializing a trivially empty Pod is:

```
{
    "spec": {
        "containers": null
    },
    "status": {}
}
```

There is no way for a Go client (structured) to produce a Pod where the `spec` and `status` keys are not  present in JSON.  Even if you had a client which can represent that, there is no way for us (in the apiserver) to detect the absence of those keys unless we 1) implement `omitzero`, 2) convert BACK to JSON to check presence/absence.  Many (most?) of our clients are go, so getting omitzero would bring new clients sort of back to par, but not fix exisiting clients.  On the server-side there are defaults, so `spec` will ALWAYS be present with non-zeroval fields.

If you look at these tags from a per-operation POV, "required" and "optional" can vary based on operation.  IMO, that is clearly NOT what the tags mean (or else we need to define which operation has which requiredness).

If you look at it from a schema POV, the tags describe the data - not the operational payload but the expected "final result" of every operation.  In that case, is it ever correct to have a Pod which doesn't have a `spec`?  You might send a serialized object which has no `spec` key, but either that will fail or the spec will get fixed server-side (e.g. sending to /status).  Doesn't that make the `spec` field required?

thockin replied:
```
# invalid spec works
$ k --context=diy get deploy hostnames -o json \
    | jq '.spec.replicas=-1' \
    | k --context=diy replace --subresource=status -f- -o kyaml

# missing spec works
$ k --context=diy get deploy hostnames -o json \
    | jq 'del(.spec)' \
    | k --context=diy replace --subresource=status -f- -o kyaml
```

thockin replied:
Declarative validation takes this as a base principle - don't revalidate unchanged data.

That doesn't help us here, though, does it?  Whether a field is optional or required only makes sense in the context of a complete object. 

One can imagine a bright future where we don't serialize empty non-pointer structs, even if I don't see us arriving there any time soon.  In the meantime, we are no closer to defining what optional/required means for non-ppointer structs :)

aaron-prindle replied:
RE: future migration of tags from +[optional|required] to +k8s:[optional|required] for openapi information:

Currently declarative validation actively forbids placing +k8s:optional, +k8s:required, or +k8s:forbidden tags on non-pointer struct fields. Attempting to do so will cause the code generator to fail with an error.

https://github.com/kubernetes/kubernetes/blob/3ec2d82da57f2018fc76ae13da7a23a3f9a364f5/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go#L103-L108

thockin replied:
+1

JoelSpeed replied:
To close this loop, we ended up writing a `nonpointerstructs` loop based on conversations we had in atlanta, that is now implemented and enabled as part of this PR, and agrees that this field remains optional

thockin reviewed (commented):
Before we go too far with this, we need to clarify non-pointer structs

JoelSpeed committed:
Update generated code

seans3 commented:
/triage accepted

JoelSpeed committed:
Enable nonpointerstructs alongside optionalorrequired

JoelSpeed committed:
Add exception for union type in scheduling API

We need to add a marker to allow podgroup to be marked up correctly as a union type, and then teach the nonpointerstructs linter how to understand this

liggitt commented:
I'm convinced by the argument that if it is impossible to create the object from a manifest which doesn't contain `spec` (or some other field), then `spec` (or that other field/stanza) is reasonable to list as required in openapi. @thockin?

JoelSpeed committed:
Update exceptions to drop "it must be optional"

JoelSpeed committed:
Update openapi/swagger

JoelSpeed committed:
Restore omitempty on ServerStorageVersion fields
